testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketReadPendingTest.java

```diff
@@ -0,0 +1,200 @@
+
+            // 4 bytes means 2 read loops for TestNumReadsRecvByteBufAllocator
+            assertTrue(serverInitializer.channelInitLatch.await(5, TimeUnit.SECONDS));
+            serverInitializer.channel.writeAndFlush(Unpooled.wrappedBuffer(new byte[4]));
+
+            serverInitializer.channel.read();
+            serverInitializer.readPendingHandler.assertAllRead();
+
+            clientChannel.read();
+            clientInitializer.readPendingHandler.assertAllRead();
+        } finally {
+            if (serverChannel != null) {
+                serverChannel.close().syncUninterruptibly();
+            }
+            if (clientChannel != null) {
+                clientChannel.close().syncUninterruptibly();
+            }
+        }
+    }
+
+    private static class ReadPendingInitializer extends ChannelInitializer<Channel> {
+        final ReadPendingReadHandler readPendingHandler = new ReadPendingReadHandler();
+        final CountDownLatch channelInitLatch = new CountDownLatch(1);
+        volatile Channel channel;
+
+        @Override
+        protected void initChannel(Channel ch) throws Exception {
+            channel = ch;
+            ch.pipeline().addLast(readPendingHandler);
+            channelInitLatch.countDown();
+        }
+    }
+
+    private static final class ReadPendingReadHandler extends ChannelInboundHandlerAdapter {
+        private final AtomicInteger count = new AtomicInteger();
+        private final CountDownLatch latch = new CountDownLatch(1);
+        private final CountDownLatch latch2 = new CountDownLatch(2);
+
+        @Override
+        public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+            ReferenceCountUtil.release(msg);
+            if (count.incrementAndGet() == 1) {
+                // Call read the first time, to ensure it is not reset the second time.
+                ctx.read();
+            }
+        }
+
+        @Override
+        public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
+            latch.countDown();
+            latch2.countDown();
+        }
+
+        void assertAllRead() throws InterruptedException {
+            assertTrue(latch.await(5, TimeUnit.SECONDS));
+            // We should only do 1 read loop, because we only called read() on the first channelRead.
+            assertFalse(latch2.await(1, TimeUnit.SECONDS));
+            assertEquals(2, count.get());
+        }
+    }
+
+    /**
+     * Designed to keep reading as long as autoread is enabled.
+     */
+    private static final class TestNumReadsRecvByteBufAllocator implements RecvByteBufAllocator {
+        private final int numReads;
+        TestNumReadsRecvByteBufAllocator(int numReads) {
+            this.numReads = numReads;
+        }
+
+        @Override
+        public Handle newHandle() {
+            return new Handle() {
+                private ChannelConfig config;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused "config" private field. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1068)